server/gfmd/user.c

```diff
@@ -1929,99 +1933,80 @@ gfm_server_user_auth_set_and_modify(struct peer *peer,
-	}
-	giant_unlock();
-
-	free(username);
-	free(auth_user_id_type_str);
-
-	return (gfm_server_put_reply(peer, diag, e, ""));
-}
-
-gfarm_error_t
-gfm_server_user_auth_set(struct peer *peer, int from_client, int skip)
-{
-	static const char diag[] = "GFM_PROTO_USER_AUTH_SET";
-	return (gfm_server_user_auth_set_and_modify(
-	    peer, from_client, skip, diag));
-}
-
-gfarm_error_t
-gfm_server_user_auth_modify(struct peer *peer, int from_client, int skip)
-{
-	static const char diag[] = "GFM_PROTO_USER_AUTH_MODIFY";
-	return (gfm_server_user_auth_set_and_modify(
-	    peer, from_client, skip, diag));
-}
-
-gfarm_error_t
-gfm_server_user_auth_remove(struct peer *peer, int from_client, int skip)
-{
-	gfarm_error_t e;
-	char *username;
-	char *auth_user_id_type_str;
-	enum auth_user_id_type auth_user_id_type;
-	struct user *u;
-	struct process *process;
-	int name_with_tenant;
-	static const char diag[] = "GFM_PROTO_USER_AUTH_REMOVE";
-
-	e = gfm_server_get_request(peer, diag, "ss",
-		   &username, &auth_user_id_type_str);
-	if (e != GFARM_ERR_NO_ERROR) {
-		gflog_debug(GFARM_MSG_UNFIXED,
-		    "%s request: %s", diag, gfarm_error_string(e));
-		return (e);
-	}
-	if (skip) {
-		free(username);
-		free(auth_user_id_type_str);
-		return (GFARM_ERR_NO_ERROR);
-	}
-
-	e = gfarm_auth_user_id_type_parse(auth_user_id_type_str,
+	e = gfarm_auth_user_id_type_from_name(auth_user_id_type_str,
 					  &auth_user_id_type);
 	if (e != GFARM_ERR_NO_ERROR) {
 		gflog_debug(GFARM_MSG_UNFIXED, "invalid auth_user_id_type");
-		return (e);
-	}
-
-	/* XXX FIXME too long giant lock */
-	giant_lock();
-
-	if ((e = rpc_name_with_tenant(peer, from_client,
-	    &name_with_tenant, &process, diag)) != GFARM_ERR_NO_ERROR) {
+	} else if (!from_client || user == NULL) {
+		e = GFARM_ERR_OPERATION_NOT_PERMITTED;
+		gflog_debug(GFARM_MSG_1003460, "%s: %s", diag,
+			gfarm_error_string(e));
+	} else if ((process = peer_get_process(peer)) == NULL) {
+		e = GFARM_ERR_OPERATION_NOT_PERMITTED;
+		gflog_debug(GFARM_MSG_UNFIXED, "%s (%s@%s): no process",
+			diag, peer_get_username(peer), peer_get_hostname(peer));
+	} else if ((tenant = process_get_tenant(process)) == NULL) {
+		e = GFARM_ERR_INTERNAL_ERROR;
+		gflog_error(GFARM_MSG_UNFIXED, "%s (%s@%s): no tenant: %s",
+			diag, peer_get_username(peer), peer_get_hostname(peer),
+			gfarm_error_string(e));
+	} else if (!user_is_tenant_admin(user, tenant)) {
+		e = GFARM_ERR_OPERATION_NOT_PERMITTED;
+		gflog_debug(GFARM_MSG_UNFIXED, "%s (%s@%s): %s",
+			diag, peer_get_username(peer), peer_get_hostname(peer),
+			gfarm_error_string(e));
+	} else if ((u = user_is_super_admin(user) ?
+			user_tenant_lookup(username) :
+			user_lookup_in_tenant(username, tenant))
+			== NULL) {
+		e = GFARM_ERR_NO_SUCH_USER;
+		gflog_debug(GFARM_MSG_UNFIXED, "%s (%s@%s) %s: %s",
+			diag, peer_get_username(peer), peer_get_hostname(peer),
+			username, gfarm_error_string(e));
+	} else if ((e = user_info_verify(&u->ui, diag)) != GFARM_ERR_NO_ERROR) {
```

Comment:
calling user_info_verify() is unnecessary.  the contents of u->ui must be already valid here.